experimental/web/sample_webgpu/main.c

```diff
@@ -279,8 +293,192 @@ static iree_status_t parse_inputs_into_call(
+    case WGPUBufferMapAsyncStatus_Error:
+      fprintf(stderr, "  buffer_map_sync_callback status: Error\n");
+      break;
+    case WGPUBufferMapAsyncStatus_DeviceLost:
+      fprintf(stderr, "  buffer_map_sync_callback status: DeviceLost\n");
+      break;
+    case WGPUBufferMapAsyncStatus_Unknown:
+    default:
+      fprintf(stderr, "  buffer_map_sync_callback status: Unknown\n");
+      break;
+  }
+
+  if (map_status != WGPUBufferMapAsyncStatus_Success) {
+    iree_allocator_free(iree_allocator_system(), userdata);
+    return;
+  }
+
+  iree_status_t status = iree_ok_status();
+
+  // TODO(scotttodd): bubble result(s) up to the caller (async + callback API)
+
+  iree_device_size_t data_length =
+      iree_hal_buffer_view_byte_length(userdata->buffer_view);
+  WGPUBuffer buffer_handle =
+      iree_hal_webgpu_buffer_handle(userdata->mappable_buffer);
+
+  // For this sample we want to print arbitrary buffers, which is easiest
+  // using the |iree_hal_buffer_view_format| function. Internally, that
+  // function requires synchronous buffer mapping, so we'll first wrap the
+  // already (async) mapped GPU memory into a heap buffer. In a less general
+  // application (or one not requiring pretty logging like this), we could
+  // skip a few buffer copies and other data transformations here.
+
+  const void* data_ptr =
+      wgpuBufferGetConstMappedRange(buffer_handle, /*offset=*/0, data_length);
+
+  // Clone into memory that isn't readonly.
+  // TODO(scotttodd): import as readonly? const void* -> void*?
+  void* data_cloned = NULL;
+  iree_host_size_t cloned_data_length =
+      iree_host_align(data_length, IREE_HAL_HEAP_BUFFER_ALIGNMENT);
+  if (iree_status_is_ok(status)) {
+    status = iree_allocator_malloc(iree_allocator_system(), cloned_data_length,
+                                   &data_cloned);
+  }
+  if (iree_status_is_ok(status)) {
+    memcpy(data_cloned, data_ptr, data_length);
+  }
+  wgpuBufferUnmap(buffer_handle);
+
+  // This looks a bit indirect, but |iree_hal_heap_buffer_wrap| isn't
+  // public while |iree_hal_allocator_import_buffer| is.
+  iree_hal_buffer_t* heap_buffer = NULL;
+  iree_hal_external_buffer_t external_buffer = {
+      .type = IREE_HAL_EXTERNAL_BUFFER_TYPE_HOST_ALLOCATION,
+      .flags = IREE_HAL_EXTERNAL_BUFFER_FLAG_NONE,
+      .size = cloned_data_length,
+      .handle = {{data_cloned}},
+  };
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_allocator_import_buffer(
+        userdata->heap_allocator,
+        (iree_hal_buffer_params_t){
+            .type = IREE_HAL_MEMORY_TYPE_HOST_LOCAL,
+            .usage = IREE_HAL_BUFFER_USAGE_MAPPING,
+            .min_alignment = IREE_HAL_HEAP_BUFFER_ALIGNMENT,
+        },
+        &external_buffer, iree_hal_buffer_release_callback_null(),
+        &heap_buffer);
+  }
+
+  // Copy the original buffer_view, backed by the cloned heap buffer instead.
+  iree_hal_buffer_view_t* buffer_view = NULL;
+  iree_hal_dim_t buffer_view_shape[10];
```

Comment:
you don't need to do this readback of the dims - iree_hal_buffer_view_shape_dims will give you direct access so you can just do `iree_hal_buffer_view_shape_rank(userdata->buffer_view), iree_hal_buffer_view_shape_dims(userdata->buffer_view),`